File: Part 1 - Data Preprocessing/Section 2 -------------------- Part 1 - Data Preprocessing --------------------/Data_Preprocessing/Hao_数据预处理.py
#DataProcessing

#import libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#import dataset
dataset = pd.read_csv('Data.csv')
#all data except purchase result
X = dataset.iloc[:,0:-1].values
#purchase result
Y = dataset.iloc[:,-1].values

#take care of missing values
from sklearn.preprocessing import Imputer
imputer = Imputer(missing_values = "NaN", strategy = 'mean', axis = 0)
imputer = imputer.fit(X[:,1:3])
X[:,1:3] = imputer.transform(X[:,1:3])

#Encoding Categorical Values
from sklearn.preprocessing import LabelEncoder
labelEncoder_X = LabelEncoder()
#create categories but put them in random order, which is not nessesary
X[:,0] = labelEncoder_X.fit_transform(X[:,0])
#create categories but put them into certain orders.
from sklearn.preprocessing import OneHotEncoder
onehotencoder = OneHotEncoder(categorical_features = [0])
Z = onehotencoder.fit_transform(X).toarray()
#change true or false to 0&1
labelEncoder_Y = LabelEncoder()
Y = labelEncoder_Y.fit_transform(Y)


#spliting the dataset into training set and testing set
from sklearn.model_selection import train_test_split
# usually 0.25 - 0.4 for testing, rest is for trainning set
# random_state = 0 will give you same random dumber everytime
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)

''' Euclidean Distance 欧式距离 is a very important relation to study '''
#Feature Scaling 特征缩放
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train_sc = sc_X.fit_transform(X_train)
'''已经拟合过了不需要再次拟合''' 
X_test_sc = sc_X.transform(X_test)

File: Part 1 - Data Preprocessing/Section 2 -------------------- Part 1 - Data Preprocessing --------------------/Data_Preprocessing/Hao_数据预处理.R
# Data Processing

# Importing DataSet
dataset = read.csv("Data.csv")

# Take care of missing data
tryout = dataset$Age
tryout1 = dataset$Age[is.na(dataset$Age)]

dataset$Age[is.na(dataset$Age)] = mean(dataset$Age, na.rm = T)
dataset$Salary[is.na(dataset$Salary)] = mean(dataset$Salary, na.rm = T)

# Transform categorical data
dataset$Country = factor(dataset$Country,
                         levels = c("France", "Germany", "Spain"),
                         labels = c(1,2,3))
dataset$Purchased = factor(dataset$Purchased,
                         levels = c("Yes", "No"),
                         labels = c(0,1))

# Splitting the dataset into training set and testing set
install.packages('caTools')
library(caTools)
# Generate random number
set.seed(123)
# Create split pattern based on random number generated
split = sample.split(dataset$Purchased, 0.8)
# Subset the source dataset into actual training set and testing set
training_set = subset(dataset, split == TRUE)
testing_set = subset(dataset, split == FALSE)

# Feature Scaling - 不能对factor进行scaling，目前只能对数字进行scaling
training_set[,2:3] = scale(training_set[,2:3])
testing_set[,2:3] = scale(testing_set[,2:3])


